pkg/scheduler/framework/interface.go

```diff
@@ -50,9 +50,54 @@ type NodeScore struct {
 	Score int64
 }
 
-// NodeToStatusMap contains the statuses of the Nodes where the incoming Pod was not schedulable.
-// A PostFilter plugin that uses this map should interpret absent Nodes as UnschedulableAndUnresolvable.
-type NodeToStatusMap map[string]*Status
+// NodeToStatus contains the statuses of the Nodes where the incoming Pod was not schedulable.
+type NodeToStatus struct {
+	// nodeToStatus contains specific statuses of the nodes.
+	nodeToStatus map[string]*Status
+	// absentNodesStatus defines a status for all nodes that are absent in nodeToStatus map.
+	// By default, all absent nodes are UnschedulableAndUnresolvable.
+	absentNodesStatus *Status
+}
+
+func NewDefaultNodeToStatus() *NodeToStatus {
+	return NewNodeToStatus(make(map[string]*Status), NewStatus(UnschedulableAndUnresolvable))
+}
+
+func NewNodeToStatus(nodeToStatus map[string]*Status, absentNodesStatus *Status) *NodeToStatus {
+	return &NodeToStatus{
+		nodeToStatus:      nodeToStatus,
+		absentNodesStatus: absentNodesStatus,
+	}
+}
+
+func (m *NodeToStatus) Get(nodeName string) *Status {
+	if status, ok := m.nodeToStatus[nodeName]; ok {
+		return status
+	}
+	return m.absentNodesStatus
+}
+
+func (m *NodeToStatus) Set(nodeName string, status *Status) {
+	m.nodeToStatus[nodeName] = status
+}
+
+func (m *NodeToStatus) Len() int {
+	return len(m.nodeToStatus)
+}
+
+func (m *NodeToStatus) AbsentNodesStatus() *Status {
+	return m.absentNodesStatus
+}
+
+func (m *NodeToStatus) SetAbsentNodesStatus(status *Status) {
+	m.absentNodesStatus = status
+}
```

Comment:
Do we need to worry about the case where multiple plugins set this to different values in the same scheduling cycle? If so, we may consider the absent node status to either stay the same, or become progressively less schedulable - `Unschedulable` to `UnschedulableAndUnresolvable` is fine, but not the reverse

Reply:
Only post filter plugins use `NodeToStatus` (receive it as an argument), but as I understand, they treat it as a read-only data structure. It means that `absentNodesStatus` shouldn't be overwritten in any place.

@sanposhiho @alculquicondor Am I right that the post filter plugins shouldn't modify the `NodeToStatus` map?

Reply:
`NodeToStatus` should be read-only in PostFilter from my memory, yes. 
But, this PR actually uses `SetAbsentNodesStatus` in preemption.go, right? Could other postfilter plugins want to do the same?

Reply:
`preemption.go` creates new NodeToStatus, but treats the PostFilter's arg as read-only. I think, to emphasize the desired way, we could create an interface with read-only methods and pass it to post filter plugins. Then, they would be able to create their own new maps (by instantiating NodeToStatus directly), but not break the current one (because of the interface).

```go
type NodeToStatusReader interface {
    Get(nodeName string) *Status
    Len() int
    AbsentNodesStatus() *Status
    ForEach(fn func(nodeName string, status *Status))
}
```

We could even name the interface `NodeToStatusMap` not to break the post filter plugins that doesn't use the map at all.

Reply:
Created `NodeToStatusReader` interface and `NodeToStatusMap` as its alias (not to break all the plugins as described above)

Reply:
Got it regarding  `SetAbsentNodesStatus` in preemption.go (Sorry I didn't read through the PR very well).

Regarding `NodeToStatusReader`, I know if we had come up with the idea at day 1, it'd have been a better idea to emphasize it's read-only though, 
to introduce this change now looks so big that the merit (emphasize it's read-only) doesn't overcome the demerit (huge effort is required for custom preemption developers).
I'm -1 for it.

Reply:
I think that passing the `NodeToStatusReader` interface (aliased as `NodeToStatusMap`) won't break more plugins than solution without the interface (i.e. passing the `NodeToStatus` struct directly). 

Plugins that doesn't read the map or those that only pass it e.g. to the `Preempt` method, won't break, because the type name `framework.NodeToStatusMap` will match the previous one. 

The only plugins that will require attention are those that read the map, but passing the struct directly will force the same, since all operations on map are hidden under the methods.

Reply:
> won't break more plugins than solution without the interface

Just looked at the previous commit. Okay, as you said, the previous commit also broke existing preemption plugins as much as the current version with `Reader` interface too.

Then, another angle though, how about just adding a status to PostFilter interface, rather than managing it in NodeToStatusMap in the first place?

```
// ...
// filterStatus is a status aggregated from prefilter or filter plugins status.  
// If the pod was rejected completely at PreFilter phase, filterStatus would be the same status as the one returned from RunPreFilterPlugins.
// And, if the Pod was rejected at Filter phase, it'd be Unschedulable if any Node was rejected with Unschedulable, otherwise UnschedulableAndUnresolvable.
PostFilter(ctx context.Context, state *CycleState, pod *v1.Pod, filteredNodeStatusMap NodeToStatusMap, filterStatus *Status) (*PostFilterResult, *Status)
```

Then, preemption PostFilter plugins can just return early when filterStatus is UnschedulableAndUnresolvable.
Would this idea reduce a required change + eliminate all the change to NodeToStatusMap? 
We, either way, break-change PostFilter interface with a new param though, people can just ignore a new param if they don't need it.

Reply:
> Then, another angle though, how about just adding a status to PostFilter interface, rather than managing it in NodeToStatusMap in the first place?

I'm not sure if this will give us any noticeable advantage. It would still require the same changes as in this PR, except modification of the map. By managing the absent status in `NodeToStatus` as proposed in this PR, it could be more straightforward to use, as both map and absent status are coupled naturally. Especially, absent status won't be accidentally ignored in the future, which could happen if handled separately. 

Your proposal would also break more PostFilter plugins, as the new arg would need to be added for each of them. In this PR `dynamicresources` plugin could not be changed at all, but did this to use the more preferable interface name over the alias.

We could even try to optimize the `NodeToStatus` in the future, by tracking changes of statuses and stating explicitly if all the nodes are `UnschedulableAndUnresolvable`, even if the map is non-empty (e.g. when filter plugins mark all the nodes as `UnschedulableAndUnresolvable`).

Reply:
Yeah, I meant, it's a wider blast radius but smaller change (my proposal) vs a narrower blast radius but bigger change (yours), and I preferr **_ed_** mine a lot more for that point.

> Especially, absent status won't be accidentally ignored in the future, which could happen if handled separately.

But, this is a critical point. Mine would require people to be aware of a new `filterStatus` for a correct preemption, which we can easily imagine people miss it.

Okay, I'm now convinced. Let's go with the current `Reader` interface design.

Reply:
@macsko Please update a release note with `Action required`.